src/store.cpp

```diff
@@ -0,0 +1,258 @@
+* with this program; if not, write to the Free Software Foundation, Inc.,
+* 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#include "otpch.h"
+
+#include "store.h"
+
+#include "pugicast.h"
+#include "database.h"
+
+StoreOffer& StoreCategory::createOffer(uint32_t offerId) {
+	offers.emplace_back(offerId);
+	return offers.back();
+}
+
+void Store::getTransactionHistory(uint32_t accountId, uint16_t page, uint32_t entriesPerPage, std::vector<StoreTransaction>& out) const {
+	Database& db = Database::getInstance();
+
+	std::ostringstream query;
+	query << "SELECT * FROM `store_history` WHERE `account_id` = " << accountId << " ORDER BY `timestamp` DESC LIMIT " << entriesPerPage << " OFFSET " << (page * entriesPerPage);
+
+	DBResult_ptr result = db.storeQuery(query.str());
+	if (result) {
+		do {
+			out.emplace_back(result->getNumber<uint32_t>("id"), result->getNumber<uint8_t>("type"), result->getNumber<int32_t>("coins"), result->getString("description"), result->getNumber<time_t>("timestamp"));
+		} while (result->next());
+	}
+}
+
+void Store::onTransactionCompleted(uint32_t accountId, uint8_t type, int32_t coins, const std::string& description) const {
+	Database& db = Database::getInstance();
+
+	std::ostringstream query;
+	query << "INSERT INTO `store_history` (`account_id`, `type`, `coins`, `description`, `timestamp`) VALUES (" << accountId << "," << type << "," << coins << "," << db.escapeString(description) << "," << time(nullptr) << ")";
+
+	db.executeQuery(query.str());
+}
+
+bool Store::loadFromXml(bool /*reloading = false*/) {
+	pugi::xml_document doc;
+	pugi::xml_parse_result result = doc.load_file("data/store/store.xml");
+	if (!result) {
+		printXMLError("Error - Store::loadFromXml", "data/store/store.xml", result);
+		return false;
+	}
+
+	loaded = true;
+
+	std::string name;
+	StoreOfferState_t state;
+	std::string description;
+	StringVector icons;
+
+	for (auto categoryNode : doc.child("store").children()) {
+		pugi::xml_attribute attr;
+
+		name.clear();
+		state = StoreOfferState_t::STORE_OFFERSTATE_NONE;
+		description.clear();
+		icons.clear();
+
+		if ((attr = categoryNode.attribute("name"))) {
+			name = attr.as_string();
+		} else {
+			std::cout << "[Error - Store::loadFromXml] Missing category name." << std::endl;
+			continue;
+		}
+
+		if ((attr = categoryNode.attribute("state"))) {
+			state = static_cast<StoreOfferState_t>(pugi::cast<uint32_t>(attr.value()));
+		}
+
+		if ((attr = categoryNode.attribute("description"))) {
+			description = attr.as_string();
+		}
+
+		if ((attr = categoryNode.attribute("icons"))) {
+			icons = explodeString(attr.as_string(), ";");
+		}
+
+		categories.emplace_back(name, state, description, icons);
+		StoreCategory& category = categories.back();
+
+		for (auto offerNode : categoryNode.children()) {
+			StoreOffer& offer = category.createOffer(runningId++);
+
+			if ((attr = offerNode.attribute("name"))) {
```

Comment:
@brunominervino Apparently travis didn't like that so add the parentheses back again, sorri :p